askama_derive/src/generator/filter.rs

```diff
@@ -0,0 +1,921 @@
+        U16 => u16,
+        U32 => u32,
+        U64 => u64,
+        U128 => u128,
+        Usize => TargetUsize;
+    }
+}
+
+struct FilterArgument {
+    name: &'static str,
+    /// If set to `None`, then a value is needed.
+    /// If set to `Some(ARGUMENT_PLACEHOLDER)`, then no value has to be assigned.
+    /// If set to `Some(&WithSpan...)`, then this value will be used if no argument was supplied.
+    default_value: Option<&'static WithSpan<'static, Expr<'static>>>,
+}
+
+/// Must be the first entry to `collect_filter_args()`'s argument `filter_args`.
+const FILTER_SOURCE: &FilterArgument = &FilterArgument {
+    name: "",
+    default_value: None,
+};
+
+const ARGUMENT_PLACEHOLDER: &WithSpan<'_, Expr<'_>> =
+    &WithSpan::new_without_span(Expr::ArgumentPlaceholder);
+
+#[inline]
+fn is_argument_placeholder(arg: &WithSpan<'_, Expr<'_>>) -> bool {
+    matches!(**arg, Expr::ArgumentPlaceholder)
+}
+
+fn no_arguments<'a, 'b>(
+    ctx: &Context<'_>,
+    name: &str,
+    args: &'b [WithSpan<'a, Expr<'a>>],
+) -> Result<&'b WithSpan<'a, Expr<'a>>, CompileError> {
+    match args {
+        [arg] => Ok(arg),
+        [_, arg, ..] => Err(ctx.generate_error(
+            format_args!("filter `{name}` does not have any arguments"),
+            arg.span(),
+        )),
+        _ => unreachable!(),
+    }
+}
+
+#[inline]
+fn collect_filter_args<'a, 'b, const N: usize>(
+    ctx: &Context<'_>,
+    name: &str,
+    node: Span<'_>,
+    input_args: &'b [WithSpan<'a, Expr<'a>>],
+    filter_args: &'static [&'static FilterArgument; N],
+) -> Result<[&'b WithSpan<'a, Expr<'a>>; N], CompileError> {
+    let mut collected_args = [ARGUMENT_PLACEHOLDER; N];
+    // rationale: less code duplication by implementing the bulk of the function non-generic
+    collect_filter_args_inner(
+        ctx,
+        name,
+        node,
+        input_args,
+        filter_args,
+        &mut collected_args,
+    )?;
+    Ok(collected_args)
+}
+
+fn collect_filter_args_inner<'a, 'b>(
+    ctx: &Context<'_>,
+    name: &str,
+    node: Span<'_>,
+    input_args: &'b [WithSpan<'a, Expr<'a>>],
+    filter_args: &'static [&'static FilterArgument],
+    collected_args: &mut [&'b WithSpan<'a, Expr<'a>>],
+) -> Result<(), CompileError> {
+    // invariant: the parser ensures that named arguments come after positional arguments
+    let mut arg_idx = 0;
+    for arg in input_args {
+        let (idx, value) = if let Expr::NamedArgument(arg_name, ref value) = **arg {
+            let Some(idx) = filter_args
+                .iter()
+                .enumerate()
+                .find_map(|(idx, arg)| (arg.name == arg_name).then_some(idx))
+            else {
+                return Err(ctx.generate_error(
+                    match filter_args.len() {
+                        1 => fmt_left!("filter `{name}` does not have any arguments"),
+                        _ => fmt_right!(
+                            "filter `{name}` does not have an argument `{}`; {}",
```

Comment:
There is already a colon in `Display` impl of `ItsArgumentsAre`.

```suggestion
                            "filter `{name}` does not have an argument `{}`{}",
```

Reply:
Oops, yes.